libs/base/Control/IOExcept.idr

```diff
@@ -2,34 +2,25 @@ module Control.IOExcept
 
 -- An IO monad with exception handling
 
-data IOExcept : Type -> Type -> Type where
-     IOM : IO (Either err a) -> IOExcept err a
+record IOExcept err a where
+     constructor IOM
+     io : IO (Either err a)
 
 instance Functor (IOExcept e) where
      map f (IOM fn) = IOM (map (map f) fn)
 
 instance Applicative (IOExcept e) where
      pure x = IOM (pure (pure x))
-     (IOM f) <*> (IOM a) = IOM (do f' <- f; a' <- a
-                                   return (f' <*> a'))
+     (IOM f) <*> (IOM a) = IOM $ (<*>) <$> f <*> a
 
 instance Monad (IOExcept e) where
-     (IOM x) >>= k = IOM (do x' <- x;
-                             case x' of
-                                  Right a => let (IOM ka) = k a in
-                                                 ka
-                                  Left err => return (Left err))
+     (IOM x) >>= f = IOM $ x >>= either (pure . Left) (io . f)
```

Comment:
I guess just `(\(IOM io) => io)` instead of `io` here would work too, and avoid making `IOException` into a record.